README.md
<h1 align="center">
  AWS API Gateway Lambda Proxy Integration module
  <br>
  <br>
</h1>

<p align="center">
  <a href="https://www.npmjs.com/package/agw-lambda-proxy"><img src="https://img.shields.io/npm/v/agw-lambda-proxy.svg" alt="npm version"></a>
  <a href="https://standardjs.com"><img src="https://img.shields.io/badge/code_style-standard-brightgreen.svg" alt="Standard - JavaScript Style Guide"></a>
  <a href="https://travis-ci.org/ovatto/agw-lambda-proxy"><img src="https://travis-ci.org/ovatto/agw-lambda-proxy.svg?branch=master" alt="build status"></a>
</p>

## What is this?

Helper module for creating Lambda functions for handling AWS API Gateway Lambda
Proxy Integration calls. See [AWS docs](https://docs.aws.amazon.com/apigateway/latest/developerguide/api-gateway-create-api-as-simple-proxy-for-lambda.html) for details.

Provides common error handling and response formatting for proxy Lambda functions.

## Installation

Simply install with [npm](http://npmjs.org):

```sh
npm install agw-lambda-proxy
```

## Usage

The module exports contains a function **createHandler** that can be used for
generating the Lambda handler function. The generated function handles the
**event**, **context**, and **callback** parameters that the Lambda runtime
passes so it can be directly used as the handler for your Lambda function.

The generator function accepts two parameters: **delegate** and **options**. The
**delegate** is a mandatory parameter and must be a function. This function is
defined by the caller and can be used for processing the Lambda request - when
Lambda function is invoked the framework passes the **event** and **context**
parameters to the delegate for request processing. The return value of the
delegate function is used for generating the Lambda function response and it
should be either a promise or a direct value if asynchronous processing is not
needed. Rejections and thrown errors are automatically handled. See
[Error handling](#error-handling) for details.

The **options** parameter can be used for overriding some of the defaults in
the request and response processing, including response headers.

### Example

Assuming you've put "index.handler" as your Lambda handler value you can use
following index.js for simply returning a response with body "static string",
status code 200 and default headers (see [CORS & response headers](#cors--response-headers)
for details).

```javascript
const delegate = (event, context) => {
  return Promise.resolve('static string')
}
module.exports = {
  handler: require('agw-lambda-proxy').createHandler(delegate)
}
```

## Response formatting

The delegate function can return the response in few different ways. For simple
responses the function can return a string value. In this case the response's
body will be the returned string and default status code 200 is used. This
case is shown above.

If the return value is an object and the object contains one of the attributes
**statusCode**, **body**, or **headers** the Lambda response is generated from
these attributes and default values for missing attribute. If the **body** is
not a string value it is automatically stringified.

```javascript
const delegate = (event, context) => {
  return callSomeCreationFunction()
    .then((created) => {
      return {
        statusCode: 201,
        body: {
          id: created.id,
          message: "Object created"
        }
      }
    })
}
module.exports = {
  handler: require('agw-lambda-proxy').createHandler(delegate)
}
```

The handling for each attribute is specified in the table below:

attribute|When specified|Default value
---------|--------------|-------------
statusCode|Parsed as an integer HTTP status code. If parsing fails, default is used.|200
body|If string, returned as is. Otherwise stringified with JSON.stringify|''
headers|Properties used as HTTP response headers.|See [CORS & response headers](#cors--response-headers)

If the returned object does not have any of the attributes **statusCode**,
**body** or **headers** the object is stringified as response body and default
status code 200 and headers are used:

```javascript
const delegate = (event, context) => {
  return {
    foo: 'bar'
  }
}
module.exports = {
  handler: require('agw-lambda-proxy').createHandler(delegate)
}
```


## CORS & response headers

### Default headers

By default, the Lambda function response will specify the header
"Access-Control-Allow-Origin" with value "\*". The default behaviour can be
changed by providing an **options** object with **headers** attribute in the
handler creation:

```javascript
const defaultHeaders = {
  'Access-Control-Allow-Origin': 'https://github.com',
  'Some-Other-Default-Header': 'header value'
}
module.exports = {
  handler: require('agw-lambda-proxy').createHandler(delegate, {headers:defaultHeaders})
}
```

Note that when overwritten default headers don't contain a value for
'Access-Control-Allow-Origin' the response will not have the default CORS
header.

### Response headers

When using the object format in the delegate return value the headers can
also be specified for individual responses:

```javascript
const delegate = (event, context) => {
  return {
    body: 'headers from return value will overwrite default headers if same keys are found',
    headers: {
      overwrite: 'from response'
    }
  }
}
const defaultHeaders = {
  overwrite: 'from defaults'
}
module.exports = {
  handler: require('agw-lambda-proxy').createHandler(delegate, {headers:defaultHeaders})
}
```

In the example above the response headers will contain value "from response" for
the header "overwrite" as delegate return value headers will take precedence
over default header values.


## Error handling

The modules handles thrown errors as well as rejections from the delegate
function. These errors are automatically converted to Lamda Proxy Integration
responses with appropriate status code. The default error code is 500 but it
can be overwritten by specifying **code** attribute on the thrown error or on the
rejection:

```javascript
const delegateThatThrows = (event, context) => {
  const error = new Error('message')
  error.code = 404
  throw error // or return Promise.reject(error)
}
```

Error message will be included in the response body in the "message"
attribute. The response body contains also a direct link to CloudWatch Logs for
this request. Returning the log link can be disabled by setting value **false**
for key **cloudWatchLogLinks** in the generator options.

## License

[MIT](LICENSE). Copyright (c) [Olli Vättö](https://github.com/ovatto).


CHANGELOG.md
# Changelog for AWS API Gateway Lambda Proxy Integration module

## [1.0.4] - 2018-01-19
* Added error logging and URL encoded log links so links are directly usable from response JSON without having to remove the escapes.

## [1.0.3] - 2018-01-19
* Added CloudWatch Logs link generation for error responses. By default the
error response now has a "log" attribute that contains a direct link to the
logs of the failing request.

## [1.0.2] - 2018-01-18
* Fixed package dependencies and setup Travis build.

## [1.0.0] - 2018-01-18
* Initial release with handler function generation.


test/test.js
/* global describe it beforeEach */
const assert = require('assert')
const createHandler = require('..').createHandler

const asPromise = fn => {
  return new Promise((resolve, reject) => {
    fn((err, res) => {
      if (err) {
        return reject(err)
      }
      return resolve(res)
    })
  })
}

const defaultHeaders = {
  'Access-Control-Allow-Origin': '*'
}

const assertResponseBody = (response, expected) => {
  const parsed = JSON.parse(response.body)
  assert.deepEqual(parsed, expected, 'Response body should match with expected')
}

describe('agw-lambda-proxy', function () {
  describe('#createHandler()', function () {
    it('should throw when delegate is not a function', function () {
      assert.throws(() => {
        createHandler(null)
      }, /"delegate" must be a function/)
    })
    it('should return a function when parameters are valid', function () {
      assert(typeof createHandler(() => 'x') === 'function')
    })
  })

  describe('#createHandler() output function', function () {
    it('should return the string returned from delegate as default response "body"', function () {
      return asPromise(callback => createHandler(() => 'expected')({}, {}, callback))
        .then((response) => {
          assert.equal(response.body, 'expected', 'Response body should be the string returned from delegate function')
          assert.equal(response.statusCode, 200, 'Default status code 200 should be returned when response doesn\'t specify status code')
          assert.deepEqual(response.headers, defaultHeaders, 'Default headers should be returned when response doesn\'t specify headers')
        })
    })
    it('should return the default response when delegate response is null', function () {
      return asPromise(callback => createHandler(() => null)({}, {}, callback))
        .then((response) => {
          assert.equal(response.body, '', 'Response body should be empty when delegate does not specify body')
          assert.equal(response.statusCode, 200, 'Default status code 200 should be returned when response doesn\'t specify status code')
          assert.deepEqual(response.headers, defaultHeaders, 'Default headers should be returned when response doesn\'t specify headers')
        })
    })
    it('should return the "statusCode" from response', function () {
      return asPromise(callback => createHandler(() => ({statusCode: 123}))({}, {}, callback))
        .then((response) => {
          assert.equal(response.body, '', 'Response body should be empty when delegate does not specify body')
          assert.equal(response.statusCode, 123, 'Response "statusCode" should match with delegate response\'s "statusCode"')
          assert.deepEqual(response.headers, defaultHeaders, 'Default headers should be returned when response doesn\'t specify headers')
        })
    })
    it('should use default when "statusCode" from response is not valid', function () {
      return asPromise(callback => createHandler(() => ({statusCode: 'xxx'}))({}, {}, callback))
        .then((response) => {
          assert.equal(response.body, '', 'Response body should be empty when delegate does not specify body')
          assert.equal(response.statusCode, 200, 'Default status code 200 should be returned when response doesn\'t specify a valid status code')
          assert.deepEqual(response.headers, defaultHeaders, 'Default headers should be returned when response doesn\'t specify headers')
        })
    })
    it('should return the "body" string from response', function () {
      return asPromise(callback => createHandler(() => ({body: 'as string'}))({}, {}, callback))
        .then((response) => {
          assert.equal(response.body, 'as string', 'Response body should match with delegate response\'s body')
          assert.equal(response.statusCode, 200, 'Default status code 200 should be returned when response doesn\'t specify status code')
          assert.deepEqual(response.headers, defaultHeaders, 'Default headers should be returned when response doesn\'t specify headers')
        })
    })
    it('should return the "body" object from response as stringified JSON', function () {
      return asPromise(callback => createHandler(() => ({body: {foo: 'bar'}}))({}, {}, callback))
        .then((response) => {
          assert.equal(response.body, '{"foo":"bar"}', 'Response body should match with delegate response\'s body')
          assert.equal(response.statusCode, 200, 'Default status code 200 should be returned when response doesn\'t specify status code')
          assert.deepEqual(response.headers, defaultHeaders, 'Default headers should be returned when response doesn\'t specify headers')
        })
    })
    it('should return the "headers" from response as well as default headers when they\'re not overwritten', function () {
      const responseHeaders = {
        MY_HEADER: 'VALUE'
      }
      return asPromise(callback => createHandler(() => ({headers: responseHeaders}))({}, {}, callback))
        .then((response) => {
          assert.equal(response.body, '', 'Response body should be empty when delegate does not specify body')
          assert.equal(response.statusCode, 200, 'Default status code 200 should be returned when response doesn\'t specify status code')
          assert.deepEqual(response.headers, Object.assign({}, defaultHeaders, responseHeaders), 'Response should have both delegate returned headers as well as default headers')
        })
    })
    it('should overwrite default "headers" when response specifies same keys', function () {
      const responseHeaders = {
        'Access-Control-Allow-Origin': 'https://foo.com'
      }
      return asPromise(callback => createHandler(() => ({headers: responseHeaders}))({}, {}, callback))
        .then((response) => {
          assert.equal(response.body, '', 'Response body should be empty when delegate does not specify body')
          assert.equal(response.statusCode, 200, 'Default status code 200 should be returned when response doesn\'t specify status code')
          assert.deepEqual(response.headers, responseHeaders, 'Response overwrite default headers when delegate returns them')
        })
    })
    it('should return options headers when options has the key "headers"', function () {
      const optionsHeaders = {
        MY_DEFAULT_HEADER: 'value'
      }
      return asPromise(callback => createHandler(() => null, {headers: optionsHeaders})({}, {}, callback))
        .then((response) => {
          assert.equal(response.body, '', 'Response body should be empty when delegate does not specify body')
          assert.equal(response.statusCode, 200, 'Default status code 200 should be returned when response doesn\'t specify status code')
          assert.deepEqual(response.headers, optionsHeaders, '"options.headers" should be returned when response doesn\'t specify headers')
        })
    })
    it('should return the returned object as stringified body when netiher statusCode, body nor headers is specified', function () {
      return asPromise(callback => createHandler(() => ({foo: 'bar'}))({}, {}, callback))
        .then((response) => {
          assert.equal(response.body, '{"foo":"bar"}', 'Response body should the stringified delegate return object')
          assert.equal(response.statusCode, 200, 'Default status code 200 should be returned when response doesn\'t specify status code')
          assert.deepEqual(response.headers, defaultHeaders, 'Default headers should be returned when response doesn\'t specify headers')
        })
    })
    it('should accept a promise returning function as delegate', function () {
      return asPromise(callback => createHandler(() => Promise.resolve('string from promise'))({}, {}, callback))
        .then((response) => {
          assert.equal(response.body, 'string from promise', 'Response body should be the string returned from delegate function promise')
          assert.equal(response.statusCode, 200, 'Default status code 200 should be returned when response doesn\'t specify status code')
          assert.deepEqual(response.headers, defaultHeaders, 'Default headers should be returned when response doesn\'t specify headers')
        })
    })
  })

  describe('#createHandler() output function error handling', function () {
    const context = {
      awsRequestId: 'UNIQUEREQUESTID'
    }
    const options = {
      logErrors: false
    }
    beforeEach(function () {
      process.env.AWS_REGION = 'aws-region-id'
      process.env.AWS_LAMBDA_LOG_GROUP_NAME = '/aws/lambda/test-log-group'
    })

    it('Should return "500" with error message when delegate function throws', function () {
      return asPromise(callback => createHandler(() => { throw new Error('thrown error message') }, options)({}, context, callback))
        .then((response) => {
          assertResponseBody(response, {
            message: 'thrown error message',
            log: 'https://aws-region-id.console.aws.amazon.com/cloudwatch/home?region=aws-region-id#logEventViewer:group=/aws/lambda/test-log-group;filter=%22UNIQUEREQUESTID%22'
          })
          assert.equal(response.statusCode, 500, 'Default error status code should be 500')
          assert.deepEqual(response.headers, defaultHeaders, 'Default headers should be returned when response doesn\'t specify headers')
        })
    })

    it('Should return "500" with error message when delegate function is rejected', function () {
      return asPromise(callback => createHandler(() => Promise.reject(new Error('rejected error message')), options)({}, context, callback))
        .then((response) => {
          assertResponseBody(response, {
            message: 'rejected error message',
            log: 'https://aws-region-id.console.aws.amazon.com/cloudwatch/home?region=aws-region-id#logEventViewer:group=/aws/lambda/test-log-group;filter=%22UNIQUEREQUESTID%22'
          })
          assert.equal(response.statusCode, 500, 'Default error status code should be 500')
          assert.deepEqual(response.headers, defaultHeaders, 'Default headers should be returned when response doesn\'t specify headers')
        })
    })

    it('Should return error\'s "code" with error message when delegate function throws', function () {
      return asPromise(callback => createHandler(() => {
        const er = new Error('thrown not found')
        er.code = 404
        throw er
      }, options)({}, context, callback))
        .then((response) => {
          assertResponseBody(response, {
            message: 'thrown not found',
            log: 'https://aws-region-id.console.aws.amazon.com/cloudwatch/home?region=aws-region-id#logEventViewer:group=/aws/lambda/test-log-group;filter=%22UNIQUEREQUESTID%22'
          })
          assert.equal(response.statusCode, 404, 'Should use the error\'s "code" as response "statusCode"')
          assert.deepEqual(response.headers, defaultHeaders, 'Default headers should be returned when response doesn\'t specify headers')
        })
    })

    it('Should return error\'s "code" with error message when delegate function is rejected', function () {
      return asPromise(callback => createHandler(() => {
        const er = new Error('rejected not found')
        er.code = 404
        return Promise.reject(er)
      }, options)({}, context, callback))
        .then((response) => {
          assertResponseBody(response, {
            message: 'rejected not found',
            log: 'https://aws-region-id.console.aws.amazon.com/cloudwatch/home?region=aws-region-id#logEventViewer:group=/aws/lambda/test-log-group;filter=%22UNIQUEREQUESTID%22'
          })
          assert.equal(response.statusCode, 404, 'Should use the error\'s "code" as response "statusCode"')
          assert.deepEqual(response.headers, defaultHeaders, 'Default headers should be returned when response doesn\'t specify headers')
        })
    })

    it('Should not return log link when "options.cloudWatchLogLinks" is false', function () {
      return asPromise(callback => createHandler(() => {
        const er = new Error('not expecting log link')
        return Promise.reject(er)
      }, {
        cloudWatchLogLinks: false,
        logErrors: false
      })({}, context, callback))
        .then((response) => {
          assertResponseBody(response, {
            message: 'not expecting log link'
          })
          assert.equal(response.statusCode, 500, 'Default error status code should be 500')
          assert.deepEqual(response.headers, defaultHeaders, 'Default headers should be returned when response doesn\'t specify headers')
        })
    })
  })
})


index.js
const url = require('url')
const querystring = require('querystring')

const formatHeaders = (options, responseHeaders = {}) => {
  return Object.assign({}, options.headers, responseHeaders)
}

const formatBody = (body) => {
  if (!body) {
    return ''
  }
  if (typeof body === 'string') {
    return body
  }
  return JSON.stringify(body)
}

const isLambdaResponse = (response) => {
  return response &&
    typeof response === 'object' &&
    ['statusCode', 'body', 'headers'].find(f => Object.prototype.hasOwnProperty.call(response, f))
}

const DEFAULT_STATUS_CODE = 200

const formatResponse = (response, options) => {
  if (isLambdaResponse(response)) {
    return {
      statusCode: Number.parseInt(response.statusCode) || DEFAULT_STATUS_CODE,
      body: formatBody(response.body),
      headers: formatHeaders(options, response.headers)
    }
  }
  return {
    statusCode: DEFAULT_STATUS_CODE,
    body: formatBody(response),
    headers: formatHeaders(options)
  }
}

const formatFilterParameter = (requestId) => querystring.escape(`"${requestId}"`)

const createLogLink = (context) => url.format({
  protocol: 'https',
  host: `${process.env.AWS_REGION}.console.aws.amazon.com`,
  pathname: '/cloudwatch/home',
  search: `region=${process.env.AWS_REGION}`,
  hash: `logEventViewer:group=${process.env.AWS_LAMBDA_LOG_GROUP_NAME};filter=${formatFilterParameter(context.awsRequestId)}`
})

const requestErrorHandler = (context, options) => (error) => {
  if (options.logErrors) {
    console.error(error)
  }
  return {
    statusCode: error.code || 500,
    body: {
      message: error.message,
      log: options.cloudWatchLogLinks ? createLogLink(context) : undefined
    }
  }
}

const DEFAULT_OPTIONS = {
  headers: {
    'Access-Control-Allow-Origin': '*'
  },
  cloudWatchLogLinks: true,
  logErrors: true
}

const createHandler = (delegate, options = {}) => {
  if (typeof delegate !== 'function') {
    throw new Error('"delegate" must be a function')
  }
  const combinedOptions = Object.assign({}, DEFAULT_OPTIONS, options)
  return (event, context, callback) => {
    return Promise.resolve()
      .then(() => delegate(event, context))
      .catch(requestErrorHandler(context, combinedOptions))
      .then((response) => callback(null, formatResponse(response, combinedOptions)))
  }
}

module.exports = {createHandler}
